### ISSUE TITLE
Reproducibility Issue - InternVL

### ISSUE THREAD

**Issue by fuzsh** [alisalamatian1/ChartGaze#2]
Hi,  

I’m working in a similar area but in a different domain. While searching, I came across your GitHub repo—it looks really interesting and quite close to what I’m trying to achieve, though I haven’t been successful so far.  

I tried running your scripts to generate attention overlays in order to explore the capabilities of InternVL. Since InternVL1 is not accessible and you mentioned using InternVL2-4B, I tested with this model. However, I wasn’t able to reproduce the attention overlays.  

I’ve been trying to work with `InternVLForConditionalGeneration` using standard models such as [InternVL2-4B from HuggingFace](https://huggingface.co/OpenGVLab/InternVL2-4B). While following your implementation, I ran into a few issues:  

1. **Model architecture mismatch**  
   The HuggingFace model uses `InternVLChatModel`, which isn’t directly compatible with `InternVLForConditionalGeneration`. Ignoring this leads to size mismatch errors. To bypass them, I tried using `ignore_size_mismatch`, but that only masks the problem.  

2. **Processor issue**  
   The `AutoProcessor` for InternVL2-4B doesn’t allow the `image` parameter—it only accepts text inputs. I’m referring to [this line of your code](https://github.com/alisalamatian1/ChartGaze/blob/2dc71e48e104bb854aff435c54fc9b00efbe5f09/models/InternVL/models/internvl_wrapper.py#L97).  
   To work around this, I replaced it with `AutoFeatureExtractor`, which made it partially functional.  

3. **Image tokens mismatch**  
   Afterwards, I encountered the error:  `Image features and image tokens do not match: tokens: 0, features: 64` This happens because tokenization sometimes requires adding image tokens. I attempted to fix this by explicitly adding them, but the issue persists.  

4. **Missing cross-attention**  
Finally, the error indicates there’s no cross-attention defined in [this line of the wrapper](https://github.com/alisalamatian1/ChartGaze/blob/2dc71e48e104bb854aff435c54fc9b00efbe5f09/models/InternVL/models/internvl_wrapper.py#L67).  

I really appreciate your work and the idea behind it. If possible, could you share more details or guidance on how to correctly reproduce the attention overlays?  

Thanks!

**Comment by giuseppecartella** [comment]
I have run into the same problems

**Comment by fuzsh** [comment]
Hi @giuseppecartella, 

I tried to modify and write another script to run the attention overlay, but I'm unsure about the attention calculation section. The section after `all_attentions = outputs.attentions` needs to be revised or rechecked. Also, I'm glad if you or other guys have ideas about whether the code is correct or not.

I used `OpenGVLab/InternVL3_5-4B-HF`, the code is as follows: 

```python
import math
import torch

import matplotlib.pyplot as plt
import torch.nn.functional as F

from PIL import Image
from torchvision import transforms
from transformers import AutoProcessor, InternVLForConditionalGeneration


def preprocess_image(pil_image, image_size=224):
    transform = transforms.Compose([
        transforms.Resize((image_size, image_size)),
        transforms.ToTensor(),
    ])
    return transform(pil_image)

main_image = Image.open('./image.png').convert('RGB')

image_tensor = preprocess_image(main_image, image_size=448)
image_pil = transforms.ToPILImage()(image_tensor)
image_pil.save('./image_preprocessed.png')

image = Image.open('./image_preprocessed.png').convert('RGB')


device = torch.device("cuda" if torch.cuda.is_available() else "CPU")
model_name = 'OpenGVLab/InternVL3_5-4B-HF'

image_processor = AutoProcessor.from_pretrained(model_name)
inputs = image_processor(
    images=image,
    text="<IMG_CONTEXT>\n what do you see in the image ?",
    truncation="only_second",
    padding=True,
    return_tensors='pt'
)

input_ids = inputs["input_ids"].to(device)
attention_mask = inputs["attention_mask"].to(device)
pixel_values = inputs["pixel_values"].to(device)

model = InternVLForConditionalGeneration.from_pretrained(
    model_name, 
    torch_dtype=torch.bfloat16, 
    device_map="auto",
    attn_implementation="eager"
)

model.config.output_attentions = True
if hasattr(model, "text_model"):
    model.text_model.config.output_attentions = True
if hasattr(model, "vision_model"):
    model.vision_model.config.output_attentions = True


with torch.no_grad():
    outputs = model(
        pixel_values=pixel_values,
        input_ids=input_ids,
        attention_mask=attention_mask,
        output_attentions=True,
        attn_implementation="eager",
        return_dict=True
    )

all_attentions = outputs.attentions

IMAGE_TOKEN_INDEX = 151671
image_token_indecies = (input_ids == IMAGE_TOKEN_INDEX).nonzero(as_tuple=True)[1]

start_idx = image_token_indecies[0]
end_idx = image_token_indecies[-1]

start_idx, end_idx

n_layer_to_collect = 12
n_layer_to_collect = min(n_layer_to_collect, len(all_attentions))

attn_maps_all_layers = []
for layer_idx, layer_attn in enumerate(all_attentions[:n_layer_to_collect]):
    # layer_attn shape: (batch_size, num_heads, seq_len, seq_len)
    # Average across heads
    layer_attn_avg_head = layer_attn.mean(dim=1)
    
    # Extract attention from text tokens to vision tokens
    # Text tokens are after the image tokens (end_idx+1:)
    # Vision tokens are from start_idx to end_idx+1
    layer_attn_text_to_vis = layer_attn_avg_head[:, end_idx+1:, start_idx:end_idx+1]
    
    # Average over all text query positions
    layer_attn_vis_avg = layer_attn_text_to_vis.mean(dim=1)
    
    attn_maps_all_layers.append(layer_attn_vis_avg)

# Average across all collected layers
attention_map = torch.stack(attn_maps_all_layers, dim=0).mean(dim=0)

# Handle batch dimension
if attention_map.dim() > 1:
    attention_map = attention_map.squeeze(0)  # Remove batch dimension

print(f"Final attention map shape: {attention_map.shape}")

# Calculate grid size for spatial arrangement
num_patches = attention_map.shape[0]
grid_size = math.ceil(math.sqrt(num_patches))

print(f"Number of vision tokens: {num_patches}")
print(f"Grid size: {grid_size}x{grid_size}")

# Pad at the END, not the beginning
if grid_size * grid_size != num_patches:
    pad_size = grid_size * grid_size - num_patches
    print(f"Padding with {pad_size} zeros at the END")
    
    zero_pad = torch.zeros(pad_size, dtype=attention_map.dtype, device=attention_map.device)
    attention_map = torch.cat([attention_map, zero_pad], dim=0)

# Normalize attention map
attention_map = (attention_map - attention_map.min()) / (attention_map.max() - attention_map.min() + 1e-8)


def create_attention_visualization(original_image, attention_map, output_path):
    """Create and save attention map visualization."""
    fig, axes = plt.subplots(1, 3, figsize=(15, 5))

    # Original image
    axes[0].imshow(original_image)
    axes[0].set_title('Original Image')
    axes[0].axis('off')

    # Attention map
    attention_map = attention_map.view(grid_size, grid_size)
    attention_resized = F.interpolate(
        attention_map.unsqueeze(0).unsqueeze(0).to(torch.float32),
        size=original_image.size[::-1],
        mode='bilinear',
        align_corners=False
    ).squeeze().cpu().numpy()

    im1 = axes[1].imshow(attention_resized, cmap='hot', alpha=0.8)
    axes[1].set_title('Attention Map')
    axes[1].axis('off')
    plt.colorbar(im1, ax=axes[1])

    # Overlay
    axes[2].imshow(original_image)
    axes[2].imshow(attention_resized, cmap='hot', alpha=0.5)
    axes[2].set_title('Attention Overlay')
    axes[2].axis('off')

    plt.tight_layout()
    plt.savefig(output_path, dpi=300, bbox_inches='tight')
    plt.close()

    print(f"Attention visualization saved to: {output_path}")

create_attention_visualization(main_image, attention_map, f"attention_visualization-{model_name.split('/')[1]}.png")
```
The generated result is as follows:

<img width="4470" height="1470" alt="Image" src="https://github.com/user-attachments/assets/5b7195b6-086b-47b4-adb1-25603421f7d0" />

**Comment by giuseppecartella** [comment]
Hello @fuzsh ,
I checked the code and it seems correct.
I have a question: do you have any clue why the input text is "What do you see in the image" instead of the question related to the specific chart image?

**Comment by fuzsh** [comment]
Hi @giuseppecartella,

I just tested with other type of normalization and it seems more meaningful, I convert the `attention_map` to 2d and afterward normalize it using the following code:
```
attention_map = attention_map.view(grid_size, grid_size)
attention_map = attention_map / (attention_map.sum(dim=-1, keepdim=True) + 1e-8)
```

about your question, I just put _"What do you see in the image"_ as example. TBH, I don't know the motivation behind.

**Comment by giuseppecartella** [comment]
Ok perfect, thank you. However, I think the standard min-max normalisation is more helpful to visualise where the model focuses the attention most.

**Comment by alisalamatian1** [comment]
@fuzsh and @giuseppecartella thanks for recognizing the work and sharing your code and sorry for a late response. I'm closing this issue, if you face other issues feel free to reach out via email.

**Comment by giuseppecartella** [comment]
Hello,

I am facing another issue,
When instantiating with InternVLForConditionalGeneration.from_pretrained I get the following warning and the loading is blocked:

You are using a model of type internvl_chat to instantiate a model of type internvl. This is not supported for all configurations of models and can yield errors.

**Comment by AmirAbaskohi** [comment]
Hi Giuseppe,

I have updated the internvl wrapper to use `AutoModel`. With the newer version of transformers, that warning happens.  I hope the issue is fixed for you.

Bests,
Amirhossein
